keys/keybase.go

```diff
@@ -7,62 +7,111 @@ import (
+	"github.com/tendermint/go-crypto/keys/bip39"
+	"github.com/tendermint/go-crypto/keys/hd"
 	dbm "github.com/tendermint/tmlibs/db"
 )
 
-// dbKeybase combines encyption and storage implementation to provide
+var _ Keybase = dbKeybase{}
+
+// Language is a language to create the BIP 39 mnemonic in.
+// Currently, only english is supported though.
+// Find a list of all supported languages in the BIP 39 spec (word lists).
+type Language int
+
+const (
+	// English is the default language to create a mnemonic.
+	// It is the only supported language by this package.
+	English Language = iota
+	// Japanese is currently not supported.
+	Japanese
+	// Korean is currently not supported.
+	Korean
+	// Spanish is currently not supported.
+	Spanish
+	// ChineseSimplified is currently not supported.
+	ChineseSimplified
+	// ChineseTraditional is currently not supported.
+	ChineseTraditional
+	// French is currently not supported.
+	French
+	// Italian is currently not supported.
+	Italian
+)
+
+// dbKeybase combines encryption and storage implementation to provide
 // a full-featured key manager
 type dbKeybase struct {
-	db    dbm.DB
-	codec words.Codec
+	db dbm.DB
 }
 
-func New(db dbm.DB, codec words.Codec) dbKeybase {
+// New creates a new keybase instance using the passed DB for reading and writing keys.
+func New(db dbm.DB) Keybase {
 	return dbKeybase{
-		db:    db,
-		codec: codec,
+		db: db,
 	}
 }
 
-var _ Keybase = dbKeybase{}
-
 // CreateMnemonic generates a new key and persists it to storage, encrypted
-// using the passphrase.  It returns the generated seedphrase
-// (mnemonic) and the key Info.  It returns an error if it fails to
+// using the provided password.
+// It returns the generated mnemonic and the key Info.
+// It returns an error if it fails to
 // generate a key for the given algo type, or if another key is
 // already stored under the same name.
-func (kb dbKeybase) CreateMnemonic(name, passphrase string, algo SignAlgo) (Info, string, error) {
-	// NOTE: secret is SHA256 hashed by secp256k1 and ed25519.
-	// 16 byte secret corresponds to 12 BIP39 words.
-	// XXX: Ledgers use 24 words now - should we ?
-	secret := crypto.CRandBytes(16)
-	priv, err := generate(algo, secret)
+func (kb dbKeybase) CreateMnemonic(name string, language Language, passwd string, algo SigningAlgo) (info Info, mnemonic string, err error) {
+	if language != English {
+		return nil, "", fmt.Errorf("unsupported language: currently only english is supported")
+	}
+	if algo != Secp256k1 {
+		err = fmt.Errorf("currently only Secp256k1 are supported as required by bip39/bip44, requested %s", algo)
+		return
+	}
+
+	// default number of words (24):
+	mnemonicS, err := bip39.NewMnemonic(bip39.FreshKey)
 	if err != nil {
-		return nil, "", err
+		return
 	}
+	// TODO(ismail): we have to be careful with the separator in non-ltr languages. Ideally, our package should provide
+	// a helper function for that
+	mnemonic = strings.Join(mnemonicS, " ")
```

Comment:
The least we can do with the separators is to have them in constants, and ideally later configurable. This is the second occurrence of the literal string and a TODO, looks like a good candidate to fix before merge.